### ISSUE TITLE
Stream mode in reports download

### ISSUE THREAD

**Issue by jesusvalle** [amz-tools/amazon-sp-api#56]
Hello,

There are some sellers that in their reports (for example, the merchant listings report), have more than 1500000 products, which implies that they exceed the node memory. 

In the previous library we used with MWS there was an option to download the reports in "stream" mode, to go through the content and incorporate it progressively. Is there any way to do it with this package?

Thank you very much!

**Comment by amz-tools** [comment]
Hi @jesusvalle,

no, there is currently no way of achieving that with the package. However we would be happy if you implement it and open  a PR, sounds like a good feature for cases such as yours. 

In the meantime maybe you can increase the memory to make it work temporary?

**Comment by jesusvalle** [comment]
Thank you, we will take it into account.

About increasing the memory, we will try as soon as we can and leave feedback.

Best regards!

**Comment by danielecr** [comment]
I am interested too. @amz-tools how could it be implemented? I had a look at the code, to stream effectively the request every stage should be streamed. In order:
 * request: the code in request.js is not suitable, it is better to rewrite a new streamed-request.js module
 * crypto code is streamable
 * iconv-lite support stream
 * csvtojson support stream

to support `options.file: filename` the method should open a write stream and pipe stream into it.
Currently `download()` method return a string.

My propose is to completely change the `download()` code using stream internally, and:
 * if specified the option `returns: "stream"` a readStream is returned (async but almost immediate).
 * if specified the option `returns: "none"` and option.file is specified it resolve only after file is written and closed. (if option.file is not accessible, not specified, etc, the method throw immediately).
 * if not specified `returns` default to "string", and a string is streamed in and promise is fulfilled when the stream end.

So `download()` will works the same way (a part from sparing memory during write of options.file), and resolve with the same string as it is now. But if the option returns is specified it change its behavior.

```
download(report_document, {
    returns: "string" | "stream" | "none"
    json: true | false
   file: <pathname>
   charset: <charset>
}
```

Here "none" is usefull to synchronize and wait for file to be wrote (stream behind the scene), while "stream" is for in-house streaming code

Of course there is the `upload()` too to be streamed, here I would change only the way the feed.file will be read and the way the request is called.

**Comment by amz-tools** [comment]
Hi @danielecr,

this sounds like a very good way of achieving it! From the first look at it I wouldn't have any changes to propose. Would you be willing to give it a shot and implement the functionality yourself or is it meant as a proposal as to how we could implement it?

**Comment by danielecr** [comment]
@amz-tools I am going to monkey patching the download method, then I will use in my app, and test. I am going to share it here.
There are some trickly things about stream and promise, and catching error using a context to avoid the crash of the nodejs app. So it would be better if it is testable before merging, then the PR will be just a step away.

I start with `download()`

**Comment by danielecr** [comment]
@amz-tools I have a POC, I still need to shape the code, but surely I miss something on first sight: xml parsing
fast-xml-parser does not support stream.
Here https://github.com/NaturalIntelligence/fast-xml-parser/issues/347 is the discussion, at the end the suggestion is to use
https://www.npmjs.com/package/arraybuffer-xml-parser

I think it is ok to use fast-xml-parser for parsing error message ( `res.code != 200`), but the other feature is better served in stream way (I did not really get the point of "complex code using stream", I am cleaning things really, you will see)

**Comment by danielecr** [comment]
Hello @jesusvalle @amz-tools , here https://github.com/danielecr/stream-amazon-sp-api is the patched download method.
It is a way to test the streamed version, and possibly compare also memory usage.
But first of all it should be used to find bug and define test-cases.
I just defined a simple test, I can use secrets + action to real test report download, but I do not know how long the report is keeped.
Is this repo already tested with some known secrets?

**Comment by amz-tools** [comment]
Hi @danielecr, 

played around a bit with the code you provided. For now I can successfully receive the stream when passed in `returns: "stream"`. So this works fine, however I have a couple of suggestions. Whenever I provide `json` or `file` option I will, of course, not be able to receive a stream. Looking at it from that perspective I don't really see the benefit from offering `none` or `string` as `returns` values.

I would suggest we just add a new option called `stream` that is `false` by default and will stream the content if set to `true`. In that case we only have to decide one thing, which can be done in three different ways:

- We ignore `json` and `file` options when `stream:true` is set
- We ignore `stream` when `json` and/or `file` is set
- We throw an error when `stream` is used together with `json` and/or `file`

I think implementing it that way would make it less complicated: Either you stream it or you don't, as a basic choice.

What do you think?

**Comment by danielecr** [comment]
Thank you for testing and feedback on the patch @amz-tools , I must say I partially agree, I feel a bit uncomfortable with `returns`-way too.
But still I see something good with returning a string.
I can exclude you are talking about complex code, because that is not the problem, I assume we are talking about interface/api complexity, together with feature provided.

You wrote "Whenever I provide json or file option I will, of course, not be able to receive a stream.", I hope you did not tested this, because this is possible, if you tested and you do not have a stream then there is a bug, at least from the point of view of the coder (that's me in this case).

I thought that one can always prefer to have the last result as 'string' or as 'stream', even if the file was written by providing 'file: whatever', and also when `json: true` is provided.

In any case the pipe order is:
 - charset
 - json
 - file
 - "returns" in stream, or in string, or none

to simplify more, maybe it is better to use an helper-something, maybe a transformer-helper to plug-in, maybe as an array of transformers to have the precedence choosen by the user.

These are naturally the stream transformer, and should provide just that interface, and this would be standard and understandable, but not so easy to use.

I also see a bug, maybe, the line https://github.com/amz-tools/amazon-sp-api/blob/main/lib/SellingPartner.js#L670:
>       decrypted = this._decodeBuffer(decrypted, res.headers, options.charset);

In patched the headers is ignored: decode from what? I think it will work anyway, but is not formally correct.

The new interface:

```
await spApi.download(reportDocument, {
        transformers: ['charset:cp1252', 'json', 'file:'+targetFile, 'output:string', customstreamtransformer]
    });
```

this way the transfomers can be moved around, there are default provided, and is extendible by customstreamtransformer inserted in the wanted position.
There is a constraint in 'output:xxxx':
 - default is 'output:none'
 - it can be used only once
 - it can be 'output:string' or 'output:stream'

I understand that ` 'file:'+targetFile` looks a bit strange, but I is versatile

**Comment by amz-tools** [comment]
Hi @danielecr,

sorry for being a bit slow in responding these days. Let me try to clarify what I mean:

When I simply use `returns:'stream'` it will return a stream (the basic goal of the enhancement) which I can work with by using `.on('data')` and `.on('end')`. However when i set `json:true` the call is not returning a stream but the final JSON. This is somehow expectable, as we would probably have to know the whole response to be able to transform it to JSON. So we can't really stream the transformed JSON object (correct me if I am wrong).

Concerning your `res.headers` question, I don't know if I understood the problem with it right. It is used to extract the charset from the response headers to use the right charset for decoding and it can optionally be set by the user resulting in ignoring the one from the response headers.

To sum up, I think we should always stream everything inside the module that is possible because it will make it more performant and this shouldn't even be possible to change by the user. The user should only have the option to decide if he/she wants a stream returned or the final string.

**Comment by danielecr** [comment]
hi @amz-tools , sorry for this quick response, I need to check it exists a ndjson format (newline delimited json). But not sure, I have to elaborate and read better your points

**Comment by danielecr** [comment]
Hi @amz-tools , I am used to reason graphically when I am confused about logic, so here is the flowchart in mermaid format, below are my points. I found this proceeding by considering input constraint and output desired processing or format.
I am not sure anymore about passing an array for specify options, anyway we will find a way.
First I should be sure there are no bias, and I need your help

```mermaid
flowchart TD;
    deflate{do deflate?}-->|YES| ocharset;
    deflate-->|NO| streampipein;
    ocharset-->iformat;
    iformat{is CSV or XML?}-->|XML| oformatjson;
    iformat-->|CSV| oformatjson2;
    oformatjson[[chunk oformatjson]]-->nostream;
    oformatjson2{is oformatjson?}-->|YES| streampipein;
    oformatjson2-->|NO| ndjson;
    ndjson[[ndjson]]-->streampipein;
    nostream(nostream)-->streaminfile;
    streaminfile[[streaminfile]]-->otype;
    streampipein[[streampipeinfile]]-->otype;
    otype{otype is stream or string?}-->|string| chunkinstring;
    otype-->|stream| streamout;
    streamout-->finaloutput;
    chunkinstring-->finaloutput;

```


I need to schematize the various options (those will be test cases, I suppose),
in a flowchart.

Here there is an input constraint and some output aims.

Input constraint:
 - is CSV or XML?
This the big question, just after deflate or ocharset.


The aims:
 - do deflate? if one want to deflate the input file or not
 - ocharset: desired output charset. It is a valid processing if 
 - oformatjson: want json format?
 - write a file: streaminfile and streampipeinfile are actions that both returns a stream
 - otype is the output type (renaming `returns`)

At the end we have a finaloutput.

My points of view are:
 - otype (aka `returns`) is always selectable.
 - streaminfile is always possible if desired.
 - using stream to write file is better in anycase (but this is just for code simplicity)
 - oformatjson: well this is a new awareness, we need here 2 kind of xml2json library (I need to search)
 - there is a big input constraint: input is in CSV or is in XML format?

And finally:
 - the only option that exclude some postprocess is do deflate (unzip?) yes or not?

**Comment by amz-tools** [comment]
Hi @danielecr,

I find it a bit hard to give some qualified input without coding and testing it on our own. I am a fan of keeping it as simple as this:

-  We stream everything internally whenever possible, no matter if the user wants a stream returned or not
- We return a stream to the user if he explicitly sets stream option
- We throw an error when the user sets stream option and another option that isn't compatible with returning a stream to the user

One question that comes to me: Is there a possible case where we can't return a stream to the user although the user hasn't set any other option?

**Comment by danielecr** [comment]
Sorry @amz-tools  I missed out that is of course technically impossible to stream an xml and transform to json, in stream format.

So is mandatory to use `fast-xml-parser` as dependency. This must be changed in my code, I am going to fix and send it in hour

The other path is anyway possible, tough ldjson format is almost the same as csv, I really do not understand the reason to implement such a thing, but people prefer different color for the same thing. (The ldjson format is in fact less efficient and more space consuming than csv, it is just algebrical calc, but you know, well, is a possible format)

I just propose to take away ldjson format from code, for now, until there are real request for such a thing.

**Comment by danielecr** [comment]
> Hi @danielecr,
> 
> I find it a bit hard to give some qualified input without coding and testing it on our own. I am a fan of keeping it as simple as this:
> 
> * We stream everything internally whenever possible, no matter if the user wants a stream returned or not
> * We return a stream to the user if he explicitly sets stream option
> * We throw an error when the user sets stream option and another option that isn't compatible with returning a stream to the user
> 
> One question that comes to me: Is there a possible case where we can't return a stream to the user although the user hasn't set any other option?

I absolutely agree with you. I am going to code, so that you can review it.

**Comment by danielecr** [comment]
> Sorry @amz-tools I missed out that is of course technically impossible to stream an xml and transform to json, in stream format.
> 
> So is mandatory to use `fast-xml-parser` as dependency. This must be changed in my code, I am going to fix and send it in hour
> 
> The other path is anyway possible, tough ldjson format is almost the same as csv, I really do not understand the reason to implement such a thing, but people prefer different color for the same thing. (The ldjson format is in fact less efficient and more space consuming than csv, it is just algebrical calc, but you know, well, is a possible format)
> 
> I just propose to take away ldjson format from code, for now, until there are real request for such a thing.

I reply on this myself useless comment, sorry I did not refreshed the page, evidentely. Well I am busy with my business task and I mix the updates. Just ignore this

**Comment by danielecr** [comment]
> Hi @danielecr,
...
> 
> One question that comes to me: Is there a possible case where we can't return a stream to the user although the user hasn't set any other option?

Hi, just back to work. I miss to answer this. From the flow chart is always possible to return a stream, although when user choose XML 2 JSON conversion, returning a stream is a bit no-sense, and maybe not efficient as expected:

The point is that it is not possible to stream-convert an XML, at least in theory, so there are no available library that provide such feature. In practice with a known XML schema and regular format is possible to implement an XML2JSON stream-converter with some effort, I found this in StackOverflow
https://stackoverflow.com/a/52562921/250970
and https://www.npmjs.com/package/xml-flow  https://www.npmjs.com/package/xml-stream so
https://www.npmjs.com/package/xtreamer

The example in xtreamer is simple enough to be used, so if amazon-sp-api one needs to transform from XML 2 json (s)he can install the dependency an use the stream as input.

Also considering this code (from https://github.com/amz-tools/amazon-sp-api/blob/main/lib/SellingPartner.js):
```
      if (options.json){
        // Transform content to json --> take content type from which to transform to json from result header
        try {
          if (res.headers['content-type'].includes('xml')){
            decrypted = this._xml_parser.parse(decrypted);
          } else if (res.headers['content-type'].includes('plain')){
            decrypted = await csv({
              delimiter:'\t',
              quote:'off'
            }).fromString(decrypted);
          }
        } catch(e){
          throw new CustomError({
            code:'PARSE_ERROR',
            message:'Could not parse result to JSON.',
            details:decrypted
          });
        }
      }
```
the plain (csv) case is inefficient for the same reason of XML (above I proposed NDJSON: newline delimited JSON format that chunk data line by line and export a regular json for each line, and not a big json)

**Comment by danielecr** [comment]
I want to rollback to this idea:

....
> 
> I would suggest we just add a new option called `stream` that is `false` by default and will stream the content if set to `true`. In that case we only have to decide one thing, which can be done in three different ways:
> 
> * We ignore `json` and `file` options when `stream:true` is set
> * We ignore `stream` when `json` and/or `file` is set
> * We throw an error when `stream` is used together with `json` and/or `file`
> 
> I think implementing it that way would make it less complicated: Either you stream it or you don't, as a basic choice.
> 

But first we need to consider that:
 1. csv to json is streamable (see [1]): so json and stream are partially compatible
 2. and that stream way is the better option for writing file, no matter of what the download() method user prefer as return type

So things become complicated. And the better choice (from the three above) is to throw error when *`stream` and `json` and content-type is xml`*. The last condition is given from upstream content-type, and not from the library usage.
This would make library usage a bit "unstable" from programmer point of view. But really this is the way:

```
try {
  await SellingPartner.download(reportDocument, {
      //charset:'cp1252',
      json: true,
      returnType: 'stream',
      file: targetFile
  });
} catch (error) {
  console.log(error); // print "It is not possible XML to JSON and to stream together."
}
```

and that would be printed "sometimes": when upstream content type is 'xml'

for the second point, "stream way is the better option...", I would prefer to say that stream default to true, but in fact I must confirm the need of sync staff provided by 'none' option, staying with 3 states:
1. returns stream: user may use freely
2. returns string: it is a buffered (and memory consuming) option
3. return 'none': it make sense only for 'file' option, if no file is provided the method return immediately

[1] **Note about csv.fromStream**
csvtojson is more clever than one would expect, see https://github.com/Keyang/node-csvtojson/blob/master/src/Result.ts
In fact it adds a '[' at begin then it parse line by line adding a regular json for each line, separated by commas.
So, it really stream the content.

**Comment by danielecr** [comment]
I updated the patched code here

https://github.com/starsellersworld/stream-amazon-sp-api

I changed the option from `returns` to `returnType`, still a three way option

I implemented upload also in stream, but the patching is commented

**Comment by shiowwj** [comment]
This will be soo helpful if implemented 🙏 to main. How can this be used on existing spapi setup? Refactor download function?

> I updated the patched code here
> 
> https://github.com/starsellersworld/stream-amazon-sp-api
> 
> I changed the option from `returns` to `returnType`, still a three way option
> 
> I implemented upload also in stream, but the patching is commented

**Comment by danielecr** [comment]
> This will be soo helpful if implemented 🙏 to main. How can this be used on existing spapi setup? Refactor download function?
> 
I am going to enable upload stream mode as soon I can test it by myself, for download stream, try
`npm i stream-amazon-sp-api`
We use this in production, but the more testa are welcome. Also I do not want to figure out various use cases, if nobody need xml to JSON in stream version then there Is no point to implement it

**Comment by shiowwj** [comment]
Thanks alot, managed to set it up and generate a local file. 

Am currently trying to stream this report result onto S3 instead of generating a file locally. However not sure if you can assist me, but the error is :
`[ERR_INVALID_ARG_TYPE]: The "list[0]" argument must be an instance of Buffer or Uint8Array. Received type string (`{"reportSpecification":{...)`

This is my code: 
`
let S3 = require("aws-sdk/clients/s3");

async function uploadReadableStream(stream) {
  const params = {Bucket: BUCKETNAME, Key: Key, Body: stream};
  return s3Client.upload(params).promise();
}
.....

let report = await SellingPartner.download(report_document, {
	json: false,
	// charset:'cp1252',
	returnType: "stream",
	// file: targetFile,
});

    // report.pipe(process.stdout)
    await uploadReadableStream(report)
`

piping the report promise streams correctly. So i presumed i could stream it into S3.

**Comment by danielecr** [comment]
@shiowwj nothing is wrong from your side, I forgot to repeat, that still "I implemented upload also in stream, but the patching is commented".
So, now, and if you like to test, try by using https://www.npmjs.com/package/patch-package
the line to uncomment is 203:
https://github.com/starsellersworld/stream-amazon-sp-api/blob/main/index.js#L203
just remove the comment
If it works, please send a PR by forking
Thank you, and sorry for the misunderstood

**Comment by shiowwj** [comment]
> @shiowwj nothing is wrong from your side, I forgot to repeat, that still "I implemented upload also in stream, but the patching is commented". So, now, and if you like to test, try by using https://www.npmjs.com/package/patch-package the line to uncomment is 203: https://github.com/starsellersworld/stream-amazon-sp-api/blob/main/index.js#L203 just remove the comment If it works, please send a PR by forking Thank you, and sorry for the misunderstood

Streaming works great! Thanks for this

**Comment by danielecr** [comment]
Hello. I updated the patch enabling streaming on uploading, it has no special options because it just create a stream inside the function from the input file.
It may be required to pass a stream in `feed` parameter, so having the option to upload `content`, `file` or `stream`, in order, and `stream` must be of type ReadStream, it is not a big change, just 2 line of code or so.

I do not know if users are confortable with options. For example, I use stream download in this way:

```
  let filename = 'output.xml';
  let streamOnTarget = (targetFile) => (stream) => {
    const ostream = require('fs').createWriteStream(targetFile);
    return new Promise((resolve, reject) => {
      stream.pipe(ostream).on('end', resolve).on('error', reject);
      ostream.on('end', resolve);
      stream.on('end',resolve);
    });
  };
  let stream = await spApi.download(feedDoc, {returnType: 'stream'});
  await streamOnTarget(filename)(stream);
```

I think it would be good to add example in README too, if I propose a patch for this project.

Also request-stream.js is a new module, and there is no timeout management, like it is in upstream version of request.js, I do not know if it is needed, because stream should manage timeout by other mean, i.e. in target stream, but really I do not know if just a call to, say `ostream.destroy()`, in the example code above has the effect to destroy also the writing stream. But basically I do not agree on handling network error by this kind of over-structure, see for example
https://stackoverflow.com/a/9910413/250970

I mean the code:
```
req.on('socket', function (socket) {
    socket.setTimeout(myTimeout);  
    socket.on('timeout', function() {
        req.abort();
    });
});
```

goes _under_ and not _over_ catching the network failure event from the lower layer available.

rif. #147  #148

**Comment by danielecr** [comment]
release v1.0.6 of stream-amazon-sp-api enable a really working upload by stream.
There are 2 important things at least. First from Amazon SP API, that demand to have Content-Length header setted *before* the real upload, and this means that really _the input must be a file_ with known length, and consequentially the applying a transformer stream that can change the size would be problematic.

The second element is the complexity of managing a duplex channel, that is send and receive, formally request and response. Both channel can be handled as stream, but depending from the http's method managing a channel as stream is more or less sensible. Given that I split the return type of requestStream in 2 cases:
 1. POST/GET methods return a (Promise of) ReadableStream.
 2. PUT method returns a (Promise of) object containing `req` (WritableStream), and a Promise of the response payload already digested (*)

(*) already digested means the json struct:
```
{
            body:body,
            chunks:chunks,
            statusCode:res.statusCode,
            headers:res.headers
          }
```

Because this is what the current of the code understand and expect to validate. Well, quite wrong here, the code expect to validate this for POST and GET too, but I can change this code to fall into the PUT-case by an option and then what SellingPartner code need is to `let {req, res} = await ...`, and `req.write(body); req.end();`, then `let response = await res`, and deal with response payload.

The other option is, of course, to let the 2 version of request and request-stream live together and includes both in the code.

What I would suggest for patch #148 is to be less invasive, I am going to comment on the PR for that

**Comment by danielecr** [comment]
just to note, AES encrypted block length is:
> cipherLen = (clearLen/16 + 1) * 16;
if additional IV block is used (and it is):
> cipherLen = (clearLen/16 + 2) * 16;
where `/` is the integer division.

**Comment by amz-tools** [comment]
`downloadStream()` and `downloadReportStream()` implemented in v.1.1.6. However will leave this open as we also want to implement an uploadStream.
